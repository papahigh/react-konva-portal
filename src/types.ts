import Konva from 'konva';
import { MutableRefObject, ReactNode } from 'react';
import { StageProps } from 'react-konva';

export interface PortalStageProps extends StageProps {
  portalLayerProps?: Partial<Konva.LayerConfig>;
}

export interface PortalStageContext {
  mount: (id: string, zIndex: number, children: ReactNode) => number;
  update: (id: string, key: number, zIndex: number, children: ReactNode) => void;
  unmount: (id: string, key: number) => void;
  addManager: (id: string, manager: PortalManagerRef) => void;
  removeManager: (id: string) => void;
}

export interface PortalManagerProps {
  id?: string;
}

export interface PortalManagerRef {
  handle(input: ManagerCommand | ManagerCommand[]): void;
}

export interface PortalProps {
  containerId?: string;
  zIndex?: number;
  children?: ReactNode;
}

export enum PortalState {
  WILL_MOUNT,
  DID_MOUNT,
  WILL_UNMOUNT,
}

export interface PortalMeta {
  key: number;
  zIndex?: number;
  children: ReactNode;
}

export type MountCommand = { type: 'mount'; key: number; zIndex: number; children: ReactNode };
export type UpdateCommand = { type: 'update'; key: number; zIndex: number; children: ReactNode };
export type UnmountCommand = { type: 'unmount'; key: number };
export type ManagerCommand = MountCommand | UpdateCommand | UnmountCommand;

export type CommandHandler<Command = ManagerCommand> = (cmd: Command) => void;

export type ForwardedRef<T> = ((instance: T | null) => void) | MutableRefObject<T | null> | null;
