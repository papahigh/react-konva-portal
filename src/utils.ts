import { ReactNode, useReducer } from 'react';
import { ManagerCommand, MountCommand, PortalMeta, UnmountCommand, UpdateCommand } from './types';

export const Z_INDEX = 0;
export const PORTAL_LAYER_ID = 'react-konva-portal';
export const ERROR_MESSAGE =
  'You should wrap your root component with <Stage /> from "react-konva-portal" in order for Portals to work';

export type RenderFunction = () => void;

export function useForceRender(): RenderFunction {
  const [, forceRender] = useReducer(i => i + 1, 0);
  return forceRender;
}

export function zIndexComparator(a: PortalMeta, b: PortalMeta) {
  const { zIndex: zIndexA = Z_INDEX } = a;
  const { zIndex: zIndexB = Z_INDEX } = b;
  if (zIndexA === zIndexB) return 0;
  return zIndexA > zIndexB ? 1 : -1;
}

export function mountCmd(key: number, zIndex: number, children: ReactNode): MountCommand {
  return { type: 'mount', key, zIndex, children };
}

export function updateCmd(key: number, zIndex: number, children: ReactNode): UpdateCommand {
  return { type: 'update', key, zIndex, children };
}

export function unmountCmd(key: number): UnmountCommand {
  return { type: 'unmount', key };
}

export function notUnmountByKey(key: number) {
  return function (item: ManagerCommand) {
    return item.key === key && item.type !== 'unmount';
  };
}

export function warnIfDev(...args: any[]) {
  if (process.env.NODE_ENV === 'development') {
    console.warn(...args);
  }
}
