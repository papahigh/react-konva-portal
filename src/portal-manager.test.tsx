import { mount, ReactWrapper } from 'enzyme';
import Konva from 'konva';
import React, { Component } from 'react';
import { Rect, Stage as ReactKonvaStage } from 'react-konva';

import Layer from './portal-layer';
import Manager from './portal-manager';
import Stage from './portal-stage';
import { PortalManagerRef } from './types';
import { mountCmd, unmountCmd, updateCmd } from './utils';

describe('portal-manager.tsx', () => {
  class App extends Component {
    public stageRef: ReactKonvaStage | null = null;
    public managerRef: PortalManagerRef | null = null;
    public render() {
      return (
        <Stage ref={ref => (this.stageRef = ref)}>
          <Layer id="layer-1">
            <Manager ref={ref => (this.managerRef = ref)} />
          </Layer>
        </Stage>
      );
    }
  }

  let wrapper: ReactWrapper<any, any, App> | undefined = undefined;
  let stage: Konva.Stage | null = null;
  let manager: PortalManagerRef | null = null;
  let layer: Konva.Layer | null = null;

  beforeEach(() => {
    wrapper = mount<App>(<App />);
    stage = wrapper.instance().stageRef?.getStage() as Konva.Stage;
    manager = wrapper.instance().managerRef as PortalManagerRef;
    layer = stage?.find('#layer-1')?.toArray()[0] as Konva.Layer;
  });

  it('should handle mount command', () => {
    expect(layer?.children.length).toEqual(0);

    manager?.handle(mountCmd(1, 1, <Rect name="rect-1" />));

    expect(layer?.children.length).toEqual(1);
    expect(layer?.children[0].name()).toEqual('rect-1');
  });

  it('should handle unmount command', () => {
    manager?.handle(mountCmd(1, 1, <Rect name="rect-1" />));

    expect(layer?.children.length).toEqual(1);
    expect(layer?.children[0].name()).toEqual('rect-1');

    manager?.handle(unmountCmd(1));

    expect(layer?.children.length).toEqual(0);
  });

  it('should handle update command', () => {
    manager?.handle(mountCmd(1, 1, <Rect name="rect-1" />));

    expect(layer?.children.length).toEqual(1);
    expect(layer?.children[0].name()).toEqual('rect-1');

    manager?.handle(updateCmd(1, 1, <Rect name="rect-1-updated" />));

    expect(layer?.children.length).toEqual(1);
    expect(layer?.children[0].name()).toEqual('rect-1-updated');
  });

  it('should handle batched commands', () => {
    manager?.handle([
      mountCmd(1, 1, <Rect name="rect-1" />),
      mountCmd(2, 1, <Rect name="rect-2" />),
      updateCmd(1, 1, <Rect name="rect-1-updated" />),
      unmountCmd(2), // drop content with key 2
      mountCmd(3, 1, <Rect name="rect-3" />),
      updateCmd(-200, 1, <Rect name="ignored" />), // unknown key
      unmountCmd(-100), // unknown key
    ]);

    expect(layer?.children.length).toEqual(2);
    expect(layer?.children[0].name()).toEqual('rect-1-updated');
    expect(layer?.children[1].name()).toEqual('rect-3');

    manager?.handle(updateCmd(1, 100, <Rect name="z-index-updated" />));

    expect(layer?.children.length).toEqual(2);
    expect(layer?.children[0].name()).toEqual('rect-3');
    expect(layer?.children[1].name()).toEqual('z-index-updated');
  });
});
