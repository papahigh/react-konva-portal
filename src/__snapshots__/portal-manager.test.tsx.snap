// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portal-manager.tsx should match snapshot 1`] = `<DocumentFragment />`;
